numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -4228,6 +4230,161 @@ _reload_guard(PyObject *NPY_UNUSED(self)) {
     Py_RETURN_NONE;
 }
 
+/* used internally */
+static void array_dlpack_internal_capsule_deleter(PyObject *self)
+{
+    DLManagedTensor *managed = 
+        (DLManagedTensor *)PyCapsule_GetPointer(self, NPY_DLPACK_INTERNAL_CAPSULE_NAME);
+    if (managed == NULL) {
+        return;
+    }
+    managed->deleter(managed);
```

Comment:
As above, this is unsafe:

* `deleter` may be null
* `PyCapsule_GetPointer` is unsafe to call if an exception is in flight, so `PyErr_Fetch` is needed first.

Reply:
@hameerabbasi, I don't think this one is resolved.

Reply:
Why do we need the `array_dlpack_internal_capsule_deleter`? Can't we use the `dlpack_capsule_deleter` instead?

Reply:
fixed.